plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,767 @@
+		return FALSE;
+	}
+#endif
+
+#ifdef HAVE_MMAN_H
+	/* map BARs */
+	for (guint i = 0; i < FU_BCM57XX_BAR_MAX; i++) {
+		int memfd;
+		struct stat st;
+		g_autofree gchar *fn = NULL;
+		g_autofree gchar *resfn = NULL;
+
+		/* open 64 bit resource */
+		resfn = g_strdup_printf ("resource%u", i * 2);
+		fn = g_build_filename (sysfs_path, resfn, NULL);
+		memfd = open (fn, O_RDWR | O_SYNC);
+		if (memfd < 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_FOUND,
+				     "error opening %s", fn);
+			return FALSE;
+		}
+		if (fstat (memfd, &st) < 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "could not stat %s", fn);
+			close (memfd);
+			return FALSE;
+		}
+
+		/* mmap */
+		g_debug ("mapping %s for 0x%x bytes", fn, (guint) st.st_size);
+		self->bar[i].buf = (guint8 *) mmap (0, st.st_size,
+						    PROT_READ | PROT_WRITE,
+						    MAP_SHARED, memfd, 0);
+		self->bar[i].bufsz = st.st_size;
+		close (memfd);
+		if (self->bar[i].buf == MAP_FAILED) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "cound not mmap %s: %s",
+				     fn, strerror(errno));
+			return FALSE;
+		}
+	}
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "mmap() not supported as sys/mman.h not available");
+	return FALSE;
+#endif
+
+	/* allow access to the flash */
+	if (!fu_bcm57xx_device_nvram_acquire_lock (self, error))
+		return FALSE;
+
+	/* verify we can read something simple */
+	vendev = fu_bcm57xx_device_bar_read (self,
+					     FU_BCM57XX_BAR_DEVICE,
+					     REG_DEVICE_PCI_VENDOR_DEVICE_ID);
+	if ((vendev & 0xffff0000) >> 16 != BCM_VENDOR_BROADCOM) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "invalid bar[0] VID, got %08x, expected %04xXXXX",
+			     vendev, (guint) BCM_VENDOR_BROADCOM);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_device_close (FuDevice *device, GError **error)
+{
+	FuBcm57xxDevice *self = FU_BCM57XX_DEVICE (device);
+
+	/* eth driver loaded */
+	if (self->ethtool_iface != NULL) {
+		close (self->ethtool_fd);
+		return TRUE;
+	}
+
+	/* no driver loaded */
+	if (!fu_bcm57xx_device_nvram_release_lock (self, error))
+		return FALSE;
+
+#ifdef HAVE_MMAN_H
+	/* unmap BARs */
+	for (guint i = 0; i < FU_BCM57XX_BAR_MAX; i++) {
+		if (self->bar[i].buf == NULL)
+			continue;
+		munmap (self->bar[i].buf, self->bar[i].bufsz);
+		self->bar[i].buf = NULL;
+		self->bar[i].bufsz = 0;
+	}
+#endif
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_bcm57xx_device_init (FuBcm57xxDevice *self)
+{
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_NO_GUID_MATCHING);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_INTERNAL);
```

Comment:
you can't guarantee this, could be in a PCIe enclosure

Reply:
Unlikely, but true I guess.